### ISSUE TITLE
useAsyncFn should add actions for set

### ISSUE THREAD

**Issue by lwyj123** [streamich/react-use#660]
**Is your feature request related to a problem? Please describe.**
It's hard to use `useAsyncFn` when I need to handle the error. for example, if the async function rejects an error, I can't roll back the state of the `state.error`

**Describe the solution you'd like**
expand the return
```Javascript
import * as ReactUse from 'react-use
const TestComponent = () => {
    const [state, {fetch, set}] = ReactUse.useAsyncFn(() => {
            return new Promise((resolve, reject) => {
                    setTimeout(() => { resolve("hello") }, 1000)
            })
    }, [settingKey])

    useEffect(() => {
        if(state.error) {
            set({
                error: undefined,
            })
            // Do something for rollback
        }
    }, [state.error])

    return <div>{state.loading}</div>
}
```
**Describe alternatives you've considered**

**Comment by wardoost** [comment]
You could add a `.catch()` to your promise chain and handle your error logic there, this will prevent the error to be updated in `state.error`.

**Comment by lwyj123** [comment]
@wardoost How about if I want to show the error and in some time I need to reset `state.error`?

**Comment by streamich** [comment]
@lwyj123 how do you solve your problem now?

**Comment by lwyj123** [comment]
@streamich I just write a custom hook in my repo.

**Comment by mendozammatias** [comment]
@streamich adding a reset callback may be as easy as adding this line on the hoook:

**Comment by mendozammatias** [comment]
@streamich adding a reset callback may be as easy as adding this line on the hoook:

`const reset = () => set(initialState)`

and

`return [state, callback, reset]; `

**Comment by streamich** [comment]
@mendozammatias will it work if you have in in-flight request and you call `reset`? Then the request completes after you called `reset`, will it overwrite the state?

**Comment by mendozammatias** [comment]
@streamich Yes, if you call `reset` and then the request is completed, the state will be overwritten.
I think it would be the expected behaviour.

                           |-----------------|-----------------|-----------------|--------------|---------
state:           initialState          customState1      customState1        initialState     customState2
callbacks:                                 something()        requestStart()        `reset()`       requestEnd()

**Comment by mendozammatias** [comment]
@streamich  Yes, if you call `reset` and then the request is completed, the state will be overwritten.
I think it would be the expected behaviour.
| | | | | | |
|-|-|-|-|-|-|
| __state__ | initialState | customState1 | customState1 | initialState | customState2
| __callbacks__ |  | something() | requestStart() | `reset()` | requestEnd()

**Comment by mendozammatias** [comment]
@streamich  Yes, if you call `reset` and then the request is completed, the state will be overwritten.
I think it would be the expected behaviour.

| |  |  |  |  |  |
| --- | --- | --- | --- | --- |---|
| __state__ | initialState | customState1 | customState1 | initialState | customState2
| __callbacks__ |  | something() | requestStart() | `reset()` | requestEnd()

**Comment by dlindenkreuz** [comment]
**Possible workaround**: use a `reset` parameter that immediately resolves with `undefined`.
This will basically reset `loading` and `error` states.

```typescript
const [signupState, signup] = useAsyncFn((email: string, reset = false) => {
  if (reset) {
    return Promise.resolve(undefined);
  }
  return sendSignup(email);
});
```

Differences between initial state and state after resetting (value key did not exist in initial state, shouldn't make a difference for most use cases):
```diff
{
- loading: false
+ loading: false,
+ value: undefined
}
```
